styx-scheduler-service/src/main/java/com/spotify/styx/state/QueuedStateManager.java

```diff
@@ -60,391 +59,239 @@
-  private final Object signal = new Object();
-  private AtomicInteger activeEvents = new AtomicInteger(0);
   private volatile boolean running = true;
 
   public QueuedStateManager(
       Time time,
-      Executor outputHandlerExecutor,
+      StripedExecutorService eventTransitionExecutor,
       Storage storage,
       BiConsumer<SequenceEvent, RunState> eventConsumer,
-      Executor eventConsumerExecutor) {
+      Executor eventConsumerExecutor,
+      OutputHandler outputHandler) {
     this.time = Objects.requireNonNull(time);
-    this.outputHandlerExecutor = Objects.requireNonNull(outputHandlerExecutor);
     this.storage = Objects.requireNonNull(storage);
     this.eventConsumer = Objects.requireNonNull(eventConsumer);
     this.eventConsumerExecutor = Objects.requireNonNull(eventConsumerExecutor);
-    this.dispatcherThread = new Thread(this::dispatch);
-    dispatcherThread.setName(DISPATCHER_THREAD_NAME);
-    dispatcherThread.start();
+    this.eventTransitionExecutor = Objects.requireNonNull(eventTransitionExecutor);
+    this.outputHandler = Objects.requireNonNull(outputHandler);
   }
 
   @Override
-  public void initialize(RunState runState) throws IsClosedException {
+  public CompletionStage<Void> trigger(WorkflowInstance workflowInstance, Trigger trigger)
+      throws IsClosedException {
     ensureRunning();
+    LOG.debug("Trigger {}", workflowInstance);
 
-    final WorkflowInstance workflowInstance = runState.workflowInstance();
-    if (states.containsKey(workflowInstance)) {
-      throw new AlreadyInitializedException("RunState initialization called on active instance "
-                                            + workflowInstance.toKey());
-    }
+    // TODO: optional retry on transaction conflict
 
-    final long counter;
-    try {
-      counter = storage.getLatestStoredCounter(workflowInstance).orElse(NO_EVENTS_PROCESSED);
-    } catch (IOException e) {
-      throw Throwables.propagate(e);
-    }
-
-    states.computeIfAbsent(workflowInstance, (wfi) -> new InstanceState(wfi, runState, counter + 1));
-  }
-
-  @Override
-  public void restore(RunState runState, long count) {
-    final WorkflowInstance workflowInstance = runState.workflowInstance();
-    if (states.containsKey(workflowInstance)) {
-      throw new RuntimeException("RunState initialization called on active instance "
-                                 + workflowInstance.toKey());
-    }
-
-    states.computeIfAbsent(workflowInstance, (wfi) -> new InstanceState(wfi, runState, count + 1));
+    return CompletableFuture.runAsync(() -> initialize(workflowInstance)).thenCompose((ignore) -> {
+      final Event event = Event.triggerExecution(workflowInstance, trigger);
+      try {
+        return receive(event);
+      } catch (IsClosedException isClosedException) {
+        LOG.warn("Failed to send 'triggerExecution' event", isClosedException);
+        // Best effort attempt to rollback the creation of the NEW state
```

Comment:
Earlier this PR directly wrote the `QUEUED` state. What is the benefit of first writing `NEW` ?